src/Mod/Sketcher/Gui/ViewProviderSketch.cpp

```diff
@@ -1212,45 +1212,165 @@ bool ViewProviderSketch::mouseButtonPressed(int Button, bool pressed, const SbVe
                     return true;
                 case STATUS_NONE: {
                     // A right click shouldn't change the Edit Mode
-                    if (preselection.isPreselectPointValid()) {
-                        return true;
+                    bool noSelection = true;
+                    bool selectPoint = false;
+                    bool selectLine = false;
+                    bool selectConstraint = false;
+                    bool selectMultiple = false;
+
+                    Gui::MenuItem geom;
+                    geom.setCommand("Sketcher context");
+                    // Gets a selection vector
+                    std::vector<Gui::SelectionObject> selection = Gui::Selection().getSelectionEx();
+                    // if something is selected
+                    if (selection.size() > 0) {
+                        const std::vector<std::string> SubNames = selection[0].getSubNames();
+
+                    if (SubNames.size() == 1) {
+                        noSelection = false;
+                        if (SubNames.begin()->substr(0, 4) == "Edge")
+                            selectLine = true;
+                        else if (SubNames.begin()->substr(0, 4) == "Vert")
+                            selectPoint = true;
+                        else if (SubNames.begin()->substr(0, 4) == "Cons")
+                            selectConstraint = true;
                     }
-                    else if (preselection.isPreselectCurveValid()) {
-                        return true;
+                    else if (SubNames.size() > 1) {
```

Comment:
There is no reason to separate the cases where `SubNames.size() == 1` and `SubNames.size() > 1`. The case `== 1 `is a special case of `> 1` that is supported by the for loop. I guess that if you remove the first case completely it will work directly.